File: 1001-2000/1882-process-tasks-using-servers/1882-process-tasks-using-servers.java
class Solution {

    class Server {
        int time;
        int index;
        int weight;
    }


    public int[] assignTasks(int[] servers, int[] tasks) {

        if(servers.length==1){
            return new int[tasks.length];
        }

        Comparator<Server> availableComparator = new Comparator<>() {
            public int compare(Server s1, Server s2) {
                if(s1.weight==s2.weight){
                    return s1.index - s2.index;
                }
                return s1.weight - s2.weight;
            }
        };

                Comparator<Server> unAvailableComparator = new Comparator<>() {
            public int compare(Server s1, Server s2) {
                return s1.time - s2.time;
            }
        };
        PriorityQueue<Server> available = new PriorityQueue<>(availableComparator);
        PriorityQueue<Server> busy = new PriorityQueue<>(unAvailableComparator);
        for(int i=0;i<servers.length;i++) {
            Server s = new Server();
            s.time = 0;
            s.weight = servers[i];
            s.index = i;
            available.add(s);
        }

        int[] ans = new int[tasks.length];
        int index = 0;
        int time = 0;

        while(index!=tasks.length) {
            while(busy.size()>0 && busy.peek().time<=time){
                available.add(busy.poll());
            }

            while(available.size()>0 && index<=time && index!=tasks.length) {
                Server s = available.poll();
                s.time = time+tasks[index];
                busy.add(s);
                ans[index]=s.index;
                index++;
            }

            if(available.size()>0){
                time++;
            } else {
                time = busy.peek().time;
            }
        }

        return ans;
    }
}

File: 1001-2000/1882-process-tasks-using-servers/1882-process-tasks-using-servers.cpp
class Solution {
public:
    vector<int> assignTasks(vector<int>& servers, vector<int>& tasks) {
        // free: {weight, index}
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> free;
        // busy: {finish_time, index}
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> busy;
        
        int n = servers.size();
        int m = tasks.size();
        vector<int> res(m);

        for (int i = 0; i < n; i++) {
            free.push({servers[i], i});
        }

        long long cur_time = 0; // 使用 long long 防止時間溢位
        for (int i = 0; i < m; i++) {
            // 1. 時間至少要是任務到達的時間 i
            cur_time = max(cur_time, (long long)i);

            // 2. 如果目前沒有空閒伺服器，時間必須跳到下一個伺服器完工的時間
            if (free.empty()) {
                cur_time = busy.top().first;
            }

            // 3. 把所有在當前時間已經完工的伺服器移回 free
            while (!busy.empty() && busy.top().first <= cur_time) {
                int idx = busy.top().second;
                free.push({servers[idx], idx});
                busy.pop();
            }

            // 4. 分配任務給最優伺服器
            auto [w, idx] = free.top();
            free.pop();
            res[i] = idx;
            busy.push({cur_time + tasks[i], idx});
        }

        return res;
    }
};